packages/connectors-web/src/alchemySmartWallet.ts

```diff
@@ -1,85 +1,233 @@
+    const innerEmitter = new Emitter("inner-uid");
+    const innerConnector = options.baseConnector({
+      ...config,
+      emitter: innerEmitter,
+    });
+
+    /*
+     * We'll return this useless provider when the smart wallet connector is not
+     * connected. We need to return _something_ in response to `getProvider`
+     * even when not connected, or Wagmi will skip over this connector.
+     *
+     * TODO: Should we return the same provider instance every time, with that
+     * instance being a proxy object routing to either this or the real provider
+     * if it exists? We would need to do that if Wagmi held on to a provider
+     * that it first got during initialization. It doesn't seem to do that
+     * today, but what if it changed behavior?
+     */
+    const loggedOutProvider: Provider = {
+      on: () => {},
+      removeListener: () => {},
+      request: async (requestParams) => {
+        if (requestParams.method === "eth_accounts") {
+          return [] as any;
+        }
+        const baseProvider = (await innerConnector.getProvider()) as any;
+        return baseProvider.request(requestParams);
+      },
+    };
+
+    let smartWalletClientPromise: Promise<SmartWalletClient> | undefined;
+
+    const resetState = (): void => {
+      smartWalletClientPromise = undefined;
+    };
+
+    const createClient = async (): Promise<SmartWalletClient> => {
+      const chainId = await innerConnector.getChainId();
+      const innerAccounts = await innerConnector.getAccounts();
+      if (innerAccounts.length === 0) {
+        throw new Error("No accounts found in base connector");
+      }
+      // TODO: do we always take the first account?
+      const signerAddress = innerAccounts[0];
+      const chain = getChainFromConfig(config, chainId);
+      const provider = (await innerConnector.getProvider()) as EIP1193Provider;
+      const signer = createWalletClient({
+        account: signerAddress,
+        chain,
+        transport: custom(provider),
+      });
+      const smartWalletClient = createSmartWalletClient({
+        signer,
+        transport: alchemyTransport({
+          apiKey: options.apiKey,
+          jwt: options.jwt,
+          url: options.url ?? "https://api.g.alchemy.com/v2",
+        }),
+        chain,
+      });
+      // TODO: Calling `client.getProvider()` has the side effect of
+      // initializing `client.account` some time after the provider is created.
+      // We should change that so it's explicit and awaitable. Until then,
+      // replicate the logic by hand.
+      const account = await smartWalletClient.requestAccount();
+      smartWalletClient.account = {
+        type: "json-rpc",
+        address: account.address,
+      };
+      return smartWalletClient;
+    };
+
+    const initializeSmartWalletClient = async (): Promise<void> => {
+      smartWalletClientPromise = createClient();
+      await smartWalletClientPromise;
```

Comment:
I think on every page load, setup is called and then connect is called right after by wagmi.  I think we might be calling createClient twice because of that.  The setup will initialize it and then the connect will reset it and initialize it again.

In the connect function, if we are already connected, I wonder if we can skip this part

```
resetState();
await innerConnector.connect(params);
```

then check if the chainId is correct and if not update it.

Reply:
Dang that's strange - I'm currently using MM to connect in the `connectors-web-example`.  I don't think it's a huge deal though.  If I get some time in a little I'll see if I can inspect why connect is getting called for me

Reply:
hmm, wagmi's `reconnect` action should definitely be calling `connect`: https://github.com/wevm/wagmi/blob/9651e18c3d140e37e16dcc80aabc29624344b7c5/packages/core/src/actions/reconnect.ts#L81

Reply:
After some refactoring I did see connect getting called on load. Not sure why it didn't come up before. In any case, this is resolved now because our `setup` no longer initializes the smart wallet client, only `connect` does.